Home_Work_27/HW27.cpp

```diff
@@ -0,0 +1,574 @@
+	std::cout << "Sum of all elements in vector of size 10000000" << std::endl;
+	std::cout << "Sum result " << resultOfSum << std::endl;
+	std::cout << "Time taken " << duration.count() << std::endl;
+
+
+	resultOfSum = 0;
+	begin = std::chrono::high_resolution_clock::now();
+
+	std::thread tread120(sumElements, vector100000000, 0, 6249999, std::ref(resultOfSum));
+	std::thread tread121(sumElements, vector100000000, 6250000, 12499999, std::ref(resultOfSum));
+	std::thread tread122(sumElements, vector100000000, 12500000, 18749999, std::ref(resultOfSum));
+	std::thread tread123(sumElements, vector100000000, 18750000, 24999999, std::ref(resultOfSum));
+	std::thread tread124(sumElements, vector100000000, 25000000, 31249999, std::ref(resultOfSum));
+	std::thread tread125(sumElements, vector100000000, 31250000, 37499999, std::ref(resultOfSum));
+	std::thread tread126(sumElements, vector100000000, 37500000, 43749999, std::ref(resultOfSum));
+	std::thread tread127(sumElements, vector100000000, 43750000, 49999999, std::ref(resultOfSum));
+	std::thread tread128(sumElements, vector100000000, 50000000, 56249999, std::ref(resultOfSum));
+	std::thread tread129(sumElements, vector100000000, 56250000, 62499999, std::ref(resultOfSum));
+	std::thread tread130(sumElements, vector100000000, 62500000, 68749999, std::ref(resultOfSum));
+	std::thread tread131(sumElements, vector100000000, 68750000, 74999999, std::ref(resultOfSum));
+	std::thread tread132(sumElements, vector100000000, 75000000, 81249999, std::ref(resultOfSum));
+	std::thread tread133(sumElements, vector100000000, 81250000, 87499999, std::ref(resultOfSum));
+	std::thread tread134(sumElements, vector100000000, 87500000, 93749999, std::ref(resultOfSum));
+	std::thread tread135(sumElements, vector100000000, 93750000, 99999999, std::ref(resultOfSum));
+
+	tread120.join();
+	tread121.join();
+	tread122.join();
+	tread123.join();
+	tread124.join();
+	tread125.join();
+	tread126.join();
+	tread127.join();
+	tread128.join();
+	tread129.join();
+	tread130.join();
+	tread131.join();
+	tread132.join();
+	tread133.join();
+	tread134.join();
+	tread135.join();
+
+	end = std::chrono::high_resolution_clock::now();
+	duration = std::chrono::duration_cast<std::chrono::microseconds>(end - begin);
+	std::cout << "Sum of all elements in vector of size 100000000" << std::endl;
+	std::cout << "Sum result " << resultOfSum << std::endl;
+	std::cout << "Time taken " << duration.count() << std::endl;
+}
+
+std::vector<int> generateVector(int size) 
+{
+	std::vector<int> resultingVector;
+
+	for (int i = 0; i < size; i++)
+	{
+		resultingVector.push_back(std::rand() % 5);
+	}
+
+	return resultingVector;
+}
+
+void sumElements(std::vector<int> inVector, int startIndex, int endIndex, int& sumResult) 
+{
+	int tempResult =0;
+
+	for (int i = startIndex; i <= endIndex; i++)
+	{
+		tempResult += inVector[i];
+	}
+	threadMutex.lock();
+	sumResult += tempResult;
+	threadMutex.unlock();
```

Comment:
locking mutex is wasteful. Why not store partial sums in vector and then add them all up after all workers without locking?